lib/core/src/Cardano/Wallet.hs

```diff
@@ -1171,103 +1165,37 @@ lookupTxIns ctx wid txins = db & \DBLayer{..} -> do
           $ withExceptT ErrDecodeTxNoSuchWallet
           $ withNoSuchWallet wid
           $ readCheckpoint wid
-    let st = getState cp
-    let extendAddr addr = case normalizeDelegationAddress @s @k @n st addr of
-            Just addr' -> addr'
-            Nothing -> error "knownAddresses should have valid fingerprints"
-    let f (addr, addrState, ix) =
-            ( extendAddr addr
-            , addrState
-            , ix )
-    let walletAddrs = map f $ knownAddresses st
-    pure $ map (tryGetTxOutPath cp walletAddrs) txins
+    pure $ map (\i -> (i, lookupTxIn cp i)) txins
   where
     db = ctx ^. dbLayer @IO @s @k
-    tryGetTxOutPath cp walletAddrs txin =
-        case UTxO.lookup txin (totalUTxO mempty cp) of
-            Nothing -> (txin, Nothing)
-            Just (txout@(TxOut addr _)) ->
-                let path = map (\(_, _,p) -> p) $
-                        filter (\(addr', _,_) -> addr' == addr) walletAddrs
-                in case path of
-                    [] -> (txin, Nothing)
-                    path':_ -> (txin, Just (txout, path'))
+    lookupTxIn :: Wallet s -> TxIn -> Maybe (TxOut, NonEmpty DerivationIndex)
+    lookupTxIn cp txIn = do
+        out@(TxOut addr _) <- UTxO.lookup txIn (totalUTxO mempty cp)
+        path <- fst $ isOurs addr (getState cp)
+        return (out, path)
 
 lookupTxOuts
-    :: forall ctx s k (n :: NetworkDiscriminant).
+    :: forall ctx s k .
         ( HasDBLayer IO s k ctx
-        , KnownAddresses s
         , IsOurs s Address
-        , DelegationAddress n k
-        , GetPurpose k
-        , WalletKey k
-        , GetAccount s k
-        , SoftDerivation k
-        , s ~ SeqState n k
         )
     => ctx
     -> WalletId
     -> [TxOut]
-    -> XPub
-    -> ExceptT ErrDecodeTx IO [(TxOut, Maybe (TxOut, NonEmpty DerivationIndex))]
-lookupTxOuts ctx wid txouts xpub = db & \DBLayer{..} -> do
+    -> ExceptT ErrDecodeTx IO [(TxOut, Maybe (NonEmpty DerivationIndex))]
+lookupTxOuts ctx wid txouts = db & \DBLayer{..} -> do
     cp <- mapExceptT atomically
           $ withExceptT ErrDecodeTxNoSuchWallet
           $ withNoSuchWallet wid
           $ readCheckpoint wid
-    let f (addr, addrState, ix) =
-            ( extendAddr cp addr
-            , addrState
-            , ix )
-    let walletAddrs = map f $ knownAddresses (getState cp)
-
-    --We are analyzing the next gapPool change addresses
```

Comment:
so we need this changeAddrs, otherwise source wallet will not see change address in outputs UNTIL the wallet discovers tx (so after submit). That is consequence of decoupling construction/signing/submitting and persisting pending things only during submitting